### PR TITLE
Attributes: strip/collapse whitespace for set values on selects

### PR DESCRIPTION
**Pull request by timmywil**
Fixes gh-2978

### CODE REVIEW

File: src/attributes/val.js [hunk 1]
```diff
@@ -4,7 +4,8 @@ define( [
 	"../core/init"
 ], function( jQuery, support ) {
 
-var rreturn = /\r/g;
+var rreturn = /\r/g,
+	rspaces = /\s/g;
```
Comment (gibson042) [inline src/attributes/val.js]: [HTML whitespace](https://html.spec.whatwg.org/#space-character) is a smaller set than that matched by `\s`, and this doesn't account for longer runs. I think you want `/[\x20\t\r\n\f]+/g`, matching various other regular expressions in our codebase.

Comment (timmywil) [inline src/attributes/val.js]: Ha, I was just going off your fiddle.

File: src/attributes/val.js [hunk 2]
```diff
@@ -128,7 +129,12 @@ jQuery.extend( {
 			set: function( elem, value ) {
 				var optionSet, option,
 					options = elem.options,
-					values = jQuery.makeArray( value ),
+
+					// Strip and collapse whitespace to normalize set values
+					// https://html.spec.whatwg.org/#strip-and-collapse-whitespace
+					values = jQuery.map( jQuery.makeArray( value ), function( val ) {
+						return jQuery.trim( val ).replace( rspaces, " " );
+					} ),
```
Comment (gibson042) [inline src/attributes/val.js]: I'm confused... what is the point of this? Fixing gh-2978 requires an update to `valHooks.option.get`, not to `valHooks.select.set`: https://github.com/jquery/jquery/issues/2978#issuecomment-194447965 . Specifically, strip and collapse whitespace is necessary to get the value of an option when it has no `value` attribute, and AFAICT nowhere else.

Comment (timmywil) [inline src/attributes/val.js]: The value coming in has to match the value coming out. `get` seems to be doing the right thing already.

File: test/unit/attributes.js [hunk 3]
```diff
@@ -1107,6 +1107,48 @@ QUnit.test( "val(select) after form.reset() (Bug #2551)", function( assert ) {
...
+		html += "<option value='bar" + value + "'>bar" + value + "</option>";
+		html += "<option value='te" + value + "st'>te" + value + "st</option>";
+		html += "<option value='" + value + "baz'>" + value + "baz</option>";
+	} );
+	$select.html( html );
+
+	jQuery.each( spaces, function( key, obj ) {
+		var val = obj.val || obj;
+		$select.val( "bar" + val );
+		assert.equal( $select.val(), "bar", "Value ending with space character (" + key + ") selected" );
```
Comment (gibson042) [inline test/unit/attributes.js]: This assertion is incorrect (as are the others). `$select.val()` should equal `"bar" + val`, because that is the value explicitly specified by the `value` attribute of the option element.

File: src/attributes/val.js (hunk 2 above)
Comment (timmywil) [inline src/attributes/val.js]: Note the new tests to confirm. `elem.value` does indeed return text when there is no `value` content attribute.

File: test/unit/attributes.js [hunk 4]
```diff
@@ -1523,17 +1565,22 @@ QUnit.test( "option value not trimmed when setting via parent select", function(
...
+QUnit.test( "Insignificant white space returned for $(option).val() (#14858, gh-2978)", function( assert ) {
+	assert.expect( 16 );
 
 	var val = jQuery( "<option></option>" ).val();
 	assert.equal( val.length, 0, "Empty option should have no value" );
 
-	val = jQuery( "<option>  </option>" ).val();
-	assert.equal( val.length, 0, "insignificant white-space returned for value" );
+	jQuery.each( [ " ", "\n", "\t", "\f", "\r" ], function() {
+		var val = jQuery( "<option>" + this + "</option>" ).val();
```
Comment (gibson042) [inline test/unit/attributes.js]: `this` abuse!

File: test/unit/attributes.js [hunk 5]
```diff
@@ -1523,17 +1565,22 @@ QUnit.test( "option value not trimmed when setting via parent select", function(
...
+		assert.equal( val.length, 0, "insignificant white-space returned for value" );
 
-	val = jQuery( "<option>  test  </option>" ).val();
-	assert.equal( val.length, 4, "insignificant white-space returned for value" );
+		val = jQuery( "<option>" + this + "test" + this + "</option>" ).val();
+		assert.equal( val.length, 4, "insignificant white-space returned for value" );
```
Comment (gibson042) [inline test/unit/attributes.js]: Why not `assert.equal( val, "test", … )`?

File: test/unit/attributes.js (hunk 3 above)
Comment (timmywil) [inline test/unit/attributes.js]: So, `text` should be stripped and collapsed, but `value` shouldn't? Taking out that trim might be a big break.

File: test/unit/attributes.js (hunk 5 above)
Comment (timmywil) [inline test/unit/attributes.js]: I just kept the old. It's accomplishing the same thing.

File: test/unit/attributes.js (hunk 4 above)
Comment (timmywil) [inline test/unit/attributes.js]: ?

**Comment by gibson042** [comment]
I'd like to see some more testing, both for non-HTML whitespace included in `\s` and/or `jQuery.trim()`, but more importantly proving a fix for the _real_ regression revealed by gh-2978, e.g.:
```js
assert.equal(
	jQuery( "<option value='&#13;foo&#9;bar&#10;'>foo bar</option>" ).val(),
	"\rfoo\tbar\n"
);
```

File: test/unit/attributes.js (hunk 4 above)
Comment (gibson042) [inline test/unit/attributes.js]: Everywhere else we `each`/`map`/etc. over primitive values, we refer to them by callback arguments rather than the `Object`-boxed `this`.

File: test/unit/attributes.js (hunk 3 above)
Comment (gibson042) [inline test/unit/attributes.js]: Correct. https://html.spec.whatwg.org/multipage/forms.html#concept-option-value specifies the `value` IDL attribute as the `value` content attribute, falling back on the [`text` IDL attribute](https://html.spec.whatwg.org/multipage/forms.html#dom-option-text), which is the element's content, concatenated, with whitespace stripped and collapsed.

Comment (timmywil) [inline test/unit/attributes.js]: Nvm, I see that's what we were doing before.

**Comment by timmywil** [comment]
@gibson042 The regression was not the only thing reported, but I see you're right about not stripping/collapsing the value.

File: src/attributes/val.js (hunk 2 above)
Comment (gibson042) [inline src/attributes/val.js]: To summarize, `jQuery.valHooks.select.set` should be left alone (the changes here reverted), but `jQuery.valHooks.option.get` as it exists in master (and here) has two problems:
* It is erroneously trimming `value` when derived from a content attribute.
* It is misimplementing [strip and collapse whitespace](https://html.spec.whatwg.org/multipage/infrastructure.html#strip-and-collapse-whitespace) when `value` is not derived from a content attribute.

**Comment by timmywil** [comment]
@gibson042 "non-space" whitespace tests added, and content attr/text behavior separated. This still isn't perfect as jQuery.trim is stripping more than we want, but perhaps it's not worth being that nitpicky.

**Comment by timmywil** [comment]
I was thinking we could normalize setting the value whether or not the option value was from a content attribute or text, but after the adjustments, I think we'd need a 2nd `.inArray` check, and I don't think it's worth it.

File: test/unit/attributes.js (hunk 4 above)
Comment (timmywil) [inline test/unit/attributes.js]: [Not](https://github.com/jquery/jquery/blob/master/src/serialize.js#L74) [everywhere](https://github.com/jquery/jquery/blob/master/src/attributes/val.js#L154), but I think I agree that naming what I'm after is better.

**Comment by gibson042** [comment]
> This still isn't perfect as jQuery.trim is stripping more than we want, but perhaps it's not worth being that nitpicky.

I get what you're saying, but I just checked, and think _all_ our internal uses of `jQuery.trim` should actually use HTML's "strip and collapse whitespace", because there are other latent edge case bugs ([example](https://jsfiddle.net/5Lhfgd3r/)). Note also that we're [already close](https://github.com/jquery/jquery/blob/a4d16a26abe3ea36337f7bfc4d1bfe13b893cd17/src/attributes/classes.js#L8), so taking the final step probably won't be that big a hit:
```js
var rhtmlSpace = /[\x20\t\r\n\f]+/g,
	stripAndCollapse = function( match, pos, str ) {
		return pos === 0 || pos === str.length - match.length ? "" : " ";
	};
…
// formerly jQuery.trim( value )
value.replace( rhtmlSpace, stripAndCollapse );
```
or
```js
var rhtmlSpace = /[\x20\t\r\n\f]+/g;
…
// formerly jQuery.trim( value )
(" " + value + " ").replace( rhtmlSpace, " " ).slice( 1, -1 );
```

File: test/unit/attributes.js [hunk 6]
```diff
@@ -1107,6 +1107,71 @@ QUnit.test( "val(select) after form.reset() (Bug #2551)", function( assert ) {
 	jQuery( "#kk" ).remove();
 } );
 
+QUnit.test( "select.val(space characters) (gh-2978)", function( assert ) {
```
Comment (markelog) [inline test/unit/attributes.js]: I guess the old way of writing tests stays forever, oh well :/

Comment (timmywil) [inline test/unit/attributes.js]: Well, there's the old way, the really old way, and the ancient way. I suspect you're talking about a newer way?

**Comment by timmywil** [comment]
I'd like to finish this one up today or tomorrow. @gibson042 Would you say this is good to go for now and we can address the #3003 later?

File: test/unit/attributes.js (hunk 6 above)
Comment (timmywil) [inline test/unit/attributes.js]: Ah, I remember now you're talking about #2886. I guess I just stuck with what I was used to.

**Comment by gibson042** [comment]
> @gibson042 Would you say this is good to go for now and we can address the #3003 later?

:+1:

File: test/unit/attributes.js (hunk 6 above)
Comment (markelog) [inline test/unit/attributes.js]: I'm just sad about this, you know? We write tests like we did 10 years ago, tests should be obvious, stateless, and isolated, now it is just ugly, like -
```js
assert.equal( $select.val(), "attr" + val, "Value ending with space character (" + key + ") selected (attr)" );
```
I can't isolate that, i don't understand what is `attr` or `key` and that thing depend on other actions, while violating every standard there is about testing.

Like ember project uses the same old QUnit, but they write them in much more structured way, like compare - https://github.com/emberjs/ember.js/blob/ed298ced885b2f4b4435fbccb26b24b9a24bb68f/tests/node/app-boot-test.js
with what we have here.

Comment (timmywil) [inline test/unit/attributes.js]: I guess I don't see the similarity in purpose between those `app-boot` tests and this. Looks like they have lots of helpers and they are establishing that HTML is generated as expected, or that certain callbacks are called/not called. We have some tests in ajax [that are similar](https://github.com/jquery/jquery/blob/master/test/unit/ajax.js#L40). I think it would help me if I could see how you'd like to write some of our existing tests to compare something closer to home.

Comment (timmywil) [inline test/unit/attributes.js]: That said, I'll land this for now so we can get 1.12.2/2.2.2 out, but I'm still interested in what you're proposing.

Comment (markelog) [inline test/unit/attributes.js]: Yeah, ajax looks better, i have shown that [before](https://github.com/jquery/jquery/pull/2810#discussion_r49638025), right now though, i'm low on motivation about this, since most of the team don't see as importance in it, which is a shame.

Comment (dmethvin) [inline test/unit/attributes.js]: Cleaning up the tests would be great, @markelog, I don't think anyone disagrees although we may not see the same degree of urgency. Are you looking for someone else on the team to show that it is important by taking it on? We all have limited resources to spend here. It seems like #2886 stalled waiting for you to give a good example of what a refactor would look like. If we had such an example and it were a "good project for a beginner" kind of thing we could keep a ticket open for that, but it probably will take someone with experience to do it.

As I said in #2886 if it's not a clear bug or feature, I'd prefer to *not* have an open ticket and instead put it on our wish list in the roadmap as @timmywil has been doing lately. We can create a ticket when we have a volunteer and a plan. Long lists of "it works now but wouldn't it be nice to do this one day" tickets are often demotivating, at least for me.

Comment (markelog) [inline test/unit/attributes.js]: > It seems like #2886 stalled waiting for you to give a good example of what a refactor would look like

I was waiting for the @mgol changes on removal tests for unsupported environments, that happened only very recently, i was planning to do that after the show/hide changes, which now never come.

> not see the same degree of urgency

The more tests we write the harder it is to change things, i would advise to check out some articles about tests writing technics (i can come with some if anyone would need that), because this should feel important, personally, i think tests sometimes even more important then the code itself.

All and all, i'm not trying to be preachy about these things without doing anything about it but it is kinda hard to fight for it when it seems no one sees why this should matter that much, like am tenager who is never understood by their parents or some stuff like that.
